File: app/routes/stockout_push.py
import logging
import datetime
import json
from flask import Blueprint, render_template, request, jsonify
from app.utils.rabbitmq import push_message
from .. import config

logger = logging.getLogger(__name__)

# ==================== 蓝图定义 ====================
stockout_push_bp = Blueprint('stockout_push', __name__, url_prefix='/stockout_push')


# ==================== 辅助函数 ====================

def transform_form_data(data):
    # 提取基础信息 - 支持从根级别或callbackResponse.deliveryOrder获取
    delivery_order_code = data.get('deliveryOrderCode') or data.get('callbackResponse', {}).get('deliveryOrder', {}).get('deliveryOrderCode')
    warehouse_code = data.get('warehouseCode') or data.get('callbackResponse', {}).get('deliveryOrder', {}).get('warehouseCode')
    
    # 增加日志以便调试
    logger.info(f"接收到的数据: {data}")
    logger.info(f"提取的delivery_order_code: {delivery_order_code}")
    logger.info(f"提取的warehouse_code: {warehouse_code}")
    
    # 获取当前时间
    current_time = datetime.datetime.now().strftime('%Y-%m-%d %H:%M:%S')
    
    # 构建订单行
    order_lines = []
    # 尝试从不同位置获取商品编码和数量，确保itemCode不为空
    # 1. 尝试从表单直接获取数组
    item_codes = data.get('itemCodes', [])
    actual_qtys = data.get('actualQtys', [])
    
    # 2. 如果没有，尝试从callbackResponse.orderLines获取
    if not item_codes:
        order_lines_data = data.get('callbackResponse', {}).get('orderLines', [])
        for item in order_lines_data:
            if isinstance(item, dict):
                item_codes.append(item.get('itemCode', ''))
                actual_qtys.append(item.get('actualQty', ''))
    
    # 3. 确保至少有一个商品
    if not item_codes:
        item_codes = ['']  # 提供一个空字符串，确保循环至少执行一次
        actual_qtys = ['1']  # 默认数量为1
    
    for i, code in enumerate(item_codes):
        # 获取数量，如果没有则默认为1
        qty = actual_qtys[i] if i < len(actual_qtys) else '1'
            
        order_lines.append({
            'actualQty': str(qty),  # 确保actualQty为字符串
            'inventoryType': 'ZP',
            'itemCode': code or '',  # 确保itemCode不为None
            'orderLineNo': '',  # 空字符串
            'ownerCode': 'XIER'
        })
    
    # 构建最终数据结构
    # 先创建空结构，然后合并预设值，最后设置我们的值以确保不会被覆盖
    transformed_data = {}
    callback_response = {}
    delivery_order = {}
    
    # 合并预设值
    if hasattr(config, 'STOCKOUT_PUSH_PRESET'):
        transformed_data = {**config.STOCKOUT_PUSH_PRESET}
        callback_response = {**config.STOCKOUT_PUSH_PRESET.get('callbackResponse', {})}
        delivery_order = {**config.STOCKOUT_PUSH_PRESET.get('callbackResponse', {}).get('deliveryOrder', {})}
    
    # 设置根级别字段
    transformed_data['outOrderCode'] = ''  # 根级别outOrderCode设为空字符串
    transformed_data['type'] = 2
    
    # 设置callbackResponse字段
    callback_response['apiMethodName'] = 'stockout.confirm'
    callback_response['responseClass'] = 'com.qimen.api.response.StockoutConfirmResponse'
    callback_response['version'] = '2.0'
    callback_response['outOrderCode'] = delivery_order_code  # 确保callbackResponse层级有outOrderCode
    callback_response['orderLines'] = order_lines
    
    # 设置deliveryOrder字段
    delivery_order['confirmType'] = 0
    delivery_order['deliveryOrderCode'] = delivery_order_code
    delivery_order['warehouseCode'] = warehouse_code
    delivery_order['operateTime'] = current_time
    delivery_order['orderConfirmTime'] = current_time
    delivery_order['orderType'] = 'PTCK'
    delivery_order['outBizCode'] = delivery_order_code  # 确保outBizCode与deliveryOrderCode一致
    delivery_order['ownerCode'] = 'XIER'
    delivery_order['status'] = 'DELIVERED'
    
    # 组装数据结构
    callback_response['deliveryOrder'] = delivery_order
    transformed_data['callbackResponse'] = callback_response
    
    logger.info(f"转换后的JSON数据: {json.dumps(transformed_data, ensure_ascii=False)}")
    
    return transformed_data


# ==================== 路由函数 ====================
@stockout_push_bp.route('/')
def index():
    # 出库单推送页面
    return render_template('stockout_push.html', preset=config.STOCKOUT_PUSH_PRESET)


@stockout_push_bp.route('/api/stockout_push', methods=['POST'])
def submit():
    # 出库单推送接口：推送数据到RabbitMQ
    try:
        # 1. 获取表单数据
        data = request.get_json()
        if not data:
            return jsonify({
                'status': 'error',
                'message': '请求数据为空'
            }), 400

        # 2. 转换表单数据格式
        transformed_data = transform_form_data(data)

        # 3. 验证必填字段
        delivery_order = transformed_data['callbackResponse']['deliveryOrder']
        required_fields = {
            'deliveryOrderCode': delivery_order.get('deliveryOrderCode'),
            'warehouseCode': delivery_order.get('warehouseCode')
        }

        missing_fields = [field for field, value in required_fields.items() if not value]
        if missing_fields:
            error_msg = f"缺少必填字段: {', '.join(missing_fields)}"
            logger.warning(f"出库单推送验证失败: {error_msg}")
            return jsonify({
                'status': 'error',
                'message': error_msg
            }), 400

        # 4. 推送消息到RabbitMQ
        logger.info(f"最终推送给RabbitMQ的报文: {json.dumps(transformed_data, ensure_ascii=False)}")
        logger.info(f"推送队列名称: {config.STOCKOUT_PUSH_QUEUE}")
        success = push_message(config.STOCKOUT_PUSH_QUEUE, transformed_data)
        logger.info(f"推送结果: {success}")

        if success:
            logger.info(f"出库单推送消息成功: {delivery_order.get('deliveryOrderCode')}")
            return jsonify({
                'status': 'success',
                'message': '出库单推送成功',
                'queue': config.STOCKOUT_PUSH_QUEUE,
                'delivery_order_code': delivery_order.get('deliveryOrderCode')
            }), 200
        else:
            logger.error(f"出库单推送消息失败: {delivery_order.get('deliveryOrderCode')}")
            return jsonify({
                'status': 'error',
                'message': '消息推送失败，请检查RabbitMQ连接'
            }), 500

    except ValueError as e:
        logger.error(f"参数验证错误: {str(e)}")
        return jsonify({
            'status': 'error',
            'message': f'参数格式错误: {str(e)}'
        }), 400
    except Exception as e:
        logger.error(f"出库单推送处理异常: {str(e)}")
        return jsonify({
            'status': 'error',
            'message': f'系统错误: {str(e)}'
        }), 500

File: app/static/js/stockout_push.js
/**
 * 出库单推送管理器
 * 处理出库单推送页面的动态功能、表单验证和JSON预览
 */
class StockoutPushManager {
    constructor() {
        // 初始化属性
        this.successModal = new bootstrap.Modal(document.getElementById('successModal'));
        this.errorModal = new bootstrap.Modal(document.getElementById('errorModal'));
        this.errorMessage = document.getElementById('error-message');
        this.detailCountInput = document.getElementById('detail-count');
        this.detailCountBadge = document.getElementById('detail-count-badge');
        this.detailFieldsContainer = document.getElementById('detail-fields');
        this.jsonPreview = document.getElementById('json-preview');
        this.copyJsonBtn = document.getElementById('copy-json');
        this.expandJsonBtn = document.getElementById('expand-json');
        this.resetBtn = document.getElementById('reset-btn');
        this.form = document.getElementById('stockout-form');
        this.isJsonExpanded = false;

        // 初始化页面
        this.init();
    }

    /**
     * 初始化页面
     */
    init() {
        // 生成初始明细项
        this.generateDetailFields();

        // 添加事件监听
        this.detailCountInput.addEventListener('input', this.debounce(() => this.generateDetailFields(), 300));
        this.form.addEventListener('input', this.debounce(() => this.updateJsonPreview(), 300));
        this.form.addEventListener('submit', (e) => this.handleSubmit(e));
        this.copyJsonBtn.addEventListener('click', () => this.copyJson());
        this.expandJsonBtn.addEventListener('click', () => this.toggleJsonExpansion());
        this.resetBtn.addEventListener('click', () => this.resetForm());

        // 页面加载时更新JSON预览
        this.updateJsonPreview();
    }

    /**
     * 生成商品明细字段
     */
    generateDetailFields() {
        // 获取明细数量并确保在有效范围内
        let count = parseInt(this.detailCountInput.value);
        if (isNaN(count) || count < 1) count = 1;
        if (count > 10) count = 10;
        this.detailCountInput.value = count;
        this.detailCountBadge.textContent = `${count} 项`;

        // 清空现有明细字段
        this.detailFieldsContainer.innerHTML = '';

        // 获取字段配置
        const fieldsConfig = JSON.parse(this.detailCountInput.getAttribute('data-fields'));

        // 生成新的明细字段
        for (let i = 0; i < count; i++) {
            const row = document.createElement('div');
            row.className = 'row g-4 mb-3 pb-3 border-b';
            row.dataset.index = i;

            // 明细项标题
            const titleCol = document.createElement('div');
            titleCol.className = 'col-12';
            titleCol.innerHTML = `<h6 class="text-muted mb-0">明细项 ${i + 1}</h6>`;
            row.appendChild(titleCol);

            // 为每个字段创建输入框
            fieldsConfig.forEach((field, fieldIndex) => {
                const col = document.createElement('div');
                col.className = 'col-md-6';

                const inputId = `${field.idPrefix}-${i}`;
                const inputName = `${field.namePrefix}[${i}]`;

                let inputHtml = `
                    <label for="${inputId}" class="form-label">
                        ${field.label} ${field.required ? '<span class="text-danger">*</span>' : ''}
                    </label>
                    <div class="input-group">
                        <span class="input-group-text">
                            <i class="${field.icon}"></i>
                        </span>
                        <input type="${field.type || 'text'}" class="form-control" id="${inputId}" name="${inputName}"
                               placeholder="${field.placeholder}" ${field.required ? 'required' : ''}
                               ${field.min !== undefined ? `min="${field.min}"` : ''}>
                    </div>
                `;

                col.innerHTML = inputHtml;
                row.appendChild(col);
            });

            this.detailFieldsContainer.appendChild(row);
        }
    }

    /**
     * 更新JSON预览
     */
    updateJsonPreview() {
        try {
            // 获取表单数据
            const formData = this.collectFormData();

            // 创建JSON结构
            const jsonData = this.buildJsonData(formData);

            // 格式化JSON并显示
            const formattedJson = this.isJsonExpanded ?
                JSON.stringify(jsonData, null, 2) :
                JSON.stringify(jsonData);

            this.jsonPreview.textContent = formattedJson;
        } catch (error) {
            this.jsonPreview.textContent = `// 生成JSON失败: ${error.message}`;
        }
    }

    /**
     * 收集表单数据
     */
    collectFormData() {
        const formData = {
            deliveryOrderCode: document.getElementById('deliveryOrderCode').value.trim(),
            warehouseCode: document.getElementById('warehouseCode').value.trim(),
            itemCodes: [],
            actualQtys: []
        };

        // 收集所有明细项
        const detailCount = parseInt(this.detailCountInput.value);
        for (let i = 0; i < detailCount; i++) {
            formData.itemCodes.push(document.getElementById(`itemCode-${i}`).value.trim());
            formData.actualQtys.push(parseInt(document.getElementById(`actualQty-${i}`).value) || 0);
        }

        return formData;
    }

    /**
     * 构建JSON数据
     */
    buildJsonData(formData) {
        // 获取预设参数
        const preset = window.preset || {};

        // 获取当前时间，格式化为yyyy-MM-dd HH:mm:ss
        const currentTime = this.formatDateTime(new Date());

        // 创建订单行
        const orderLines = [];
        for (let i = 0; i < formData.itemCodes.length; i++) {
            orderLines.push({
                actualQty: formData.actualQtys[i],
                inventoryType: 'ZP',
                itemCode: formData.itemCodes[i],
                orderLineNo: `${i + 1}`,
                ownerCode: 'XIER'
            });
        }

        // 构建完整JSON
        const jsonData = {
            callbackResponse: {
                apiMethodName: 'stockout.confirm',
                deliveryOrder: {
                    confirmType: 0,
                    deliveryOrderCode: formData.deliveryOrderCode,
                    operateTime: currentTime,
                    orderConfirmTime: currentTime,
                    orderType: 'PTCK',
                    outBizCode: '',
                    ownerCode: 'XIER',
                    status: 'DELIVERED',
                    warehouseCode: formData.warehouseCode
                },
                orderLines: orderLines,
                responseClass: 'com.qimen.api.response.StockoutConfirmResponse',
                version: '2.0'
            },
            outOrderCode: formData.deliveryOrderCode,
            type: 2
        };

        return jsonData;
    }

    /**
     * 格式化日期时间
     */
    formatDateTime(date) {
        const year = date.getFullYear();
        const month = String(date.getMonth() + 1).padStart(2, '0');
        const day = String(date.getDate()).padStart(2, '0');
        const hours = String(date.getHours()).padStart(2, '0');
        const minutes = String(date.getMinutes()).padStart(2, '0');
        const seconds = String(date.getSeconds()).padStart(2, '0');

        return `${year}-${month}-${day} ${hours}:${minutes}:${seconds}`;
    }

    /**
     * 处理表单提交
     */
    handleSubmit(e) {
        e.preventDefault();

        // 验证表单
        if (!this.form.checkValidity()) {
            e.stopPropagation();
            this.form.classList.add('was-validated');
            this.showToast('请填写所有必填字段', 'error');
            return;
        }

        try {
            // 获取表单数据并构建JSON
            const formData = this.collectFormData();
            const jsonData = this.buildJsonData(formData);

            // 发送数据到后端
            fetch('/stockout_push/api/stockout_push', {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json'
                },
                body: JSON.stringify(jsonData)
            })
            .then(response => {
                if (!response.ok) {
                    throw new Error('网络响应错误');
                }
                return response.json();
            })
            .then(data => {
                if (data.status === 'success') {
                    this.showToast('出库单已成功推送', 'success');
                } else {
                    this.errorMessage.textContent = data.message || '推送失败，请稍后重试';
                    this.errorModal.show();
                    this.showToast(data.message || '推送失败，请稍后重试', 'error');
                }
            })
            .catch(error => {
                this.errorMessage.textContent = `推送失败: ${error.message}`;
                this.errorModal.show();
                this.showToast(`推送失败: ${error.message}`, 'error');
            });
        } catch (error) {
            this.errorMessage.textContent = `处理请求时出错: ${error.message}`;
            this.errorModal.show();
            this.showToast(`处理请求时出错: ${error.message}`, 'error');
        }
    }

    /**
     * 复制JSON到剪贴板
     */
    copyJson() {
        const jsonText = this.jsonPreview.textContent;
        navigator.clipboard.writeText(jsonText)
            .then(() => {
                this.showToast('JSON已复制到剪贴板', 'success');
            })
            .catch(err => {
                this.showToast('复制失败，请手动复制', 'error');
            });
    }

    /**
     * 切换JSON展开/折叠状态
     */
    toggleJsonExpansion() {
        this.isJsonExpanded = !this.isJsonExpanded;
        this.expandJsonBtn.innerHTML = this.isJsonExpanded ?
            '<i class="fas fa-compress me-1"></i>折叠' :
            '<i class="fas fa-expand me-1"></i>展开';
        this.updateJsonPreview();
    }

    /**
     * 重置表单
     */
    resetForm() {
        this.form.reset();
        this.form.classList.remove('was-validated');
        this.detailCountInput.value = 1;
        this.generateDetailFields();
        this.updateJsonPreview();
        this.showToast('表单已重置', 'info');
    }

    /**
     * 防抖函数
     */
    debounce(func, wait) {
        let timeout;
        return function() {
            const context = this;
            const args = arguments;
            clearTimeout(timeout);
            timeout = setTimeout(() => func.apply(context, args), wait);
        };
    }

    /**
     * 显示提示消息
     */
    showToast(message, type = 'info') {
        // 检查是否已有toast容器
        let toastContainer = document.getElementById('toast-container');
        if (!toastContainer) {
            toastContainer = document.createElement('div');
            toastContainer.id = 'toast-container';
            toastContainer.className = 'toast-container position-fixed top-0 end-0 p-3';
            toastContainer.style.zIndex = '1055';
            document.body.appendChild(toastContainer);
        }

        // 创建toast元素
        const toastId = `toast-${Date.now()}`;
        const toastHtml = `
            <div id="${toastId}" class="toast align-items-center text-white bg-${type} border-0" role="alert" aria-live="assertive" aria-atomic="true">
                <div class="d-flex">
                    <div class="toast-body">
                        ${message}
                    </div>
                    <button type="button" class="btn-close btn-close-white me-2 m-auto" data-bs-dismiss="toast" aria-label="Close"></button>
                </div>
            </div>
        `;

        toastContainer.innerHTML += toastHtml;

        // 显示toast
        const toast = new bootstrap.Toast(document.getElementById(toastId));
        toast.show();

        // 3秒后自动移除
        setTimeout(() => {
            const element = document.getElementById(toastId);
            if (element) {
                element.remove();
            }
        }, 3000);
    }
}

// 页面加载完成后初始化
document.addEventListener('DOMContentLoaded', function() {
    new StockoutPushManager();
});